src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,183 @@
+        {
+            // Since DocumentExecuter catches and wraps most exceptions, this should only be thrown
+            // if the request was canceled or if ThrowOnUnhandledException is true. And since
+            // OperationCanceledExceptions are not a fault, we do not record them as such.
+            if (ex is not OperationCanceledException)
+            {
+#if NET6_0_OR_GREATER
+                activity.SetStatus(ActivityStatusCode.Error);
+#else
+                activity.SetTag("otel.status_code", "ERROR");
+#endif
+            }
+            _telemetryOptions.EnrichWithException(activity, ex);
+            throw;
+        }
+
+        // record the status
+        await SetResultTagsAsync(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivityAsync(ExecutionOptions options)
+        => new(ActivitySource.StartActivity(ACTIVITY_OPERATION_NAME));
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="ExecutionOptions"/>.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.document</c> tags.
+    /// </summary>
+    protected virtual Task SetInitialTagsAsync(Activity activity, ExecutionOptions options)
+    {
+        activity.SetTag("graphql.operation.name", options.OperationName);
+        if (_telemetryOptions.RecordDocument /* && activity.IsAllDataRequested */)
+            activity.SetTag("graphql.document", options.Query);
+        _telemetryOptions.EnrichWithExecutionOptions(activity, options);
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="GraphQLOperationDefinition"/> and other values.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.operation.type</c> tags, and sets the
+    /// <see cref="Activity.DisplayName"/> property based on the operation name and type.
+    /// </summary>
+    protected virtual Task SetOperationTagsAsync(Activity activity, ExecutionOptions options, ISchema schema, GraphQLDocument document, GraphQLOperationDefinition operation)
+    {
+        var operationType = operation.Operation switch
+        {
+            OperationType.Query => "query",
+            OperationType.Mutation => "mutation",
+            OperationType.Subscription => "subscription",
+            _ => "unknown", // cannot occur
+        };
+        // https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/instrumentation/graphql/
+        activity.SetTag("graphql.operation.type", operationType);
+        var operationName = operation.Name?.StringValue;
+        activity.SetTag("graphql.operation.name", operationName);
```

Comment:
Always set or only when not null ?

Reply:
Changed to only when not null.  Note that current .NET documentation says:

> If the input value is null and if the collection has any tag with the same key, then this tag will get removed from the collection.